### ISSUE TITLE
Remove as many options as possible

### ISSUE THREAD

**Issue by phated** [floridoo/gulp-sourcemaps#285]
I need to get an idea of how people are using all the different options and why they are using them.  My goal is to understand and automatically handle those operations so this plugin doesn't need so much configuration.

Please post your option usage and __why__ you use them below.  Otherwise, your use case might disappear with the option.

**Comment by phated** [comment]
`mapSources` has been deprecated with a separate stream API as the replacement

**Comment by phated** [comment]
`identityMap` has been deprecated with a separate stream API as the replacement.

**Comment by coreyfarrell** [comment]
My use case is for `mapSources` is to rewrite the source paths after source contents have been loaded.  Once `@gulp-sourcemaps/sources-content` implementation is merged I won't need this.

One comment, part of me wishes that `@gulp-sourcemaps/identity-map` were not pulled in by this module and it were up to users to just directly use it.  Mostly I'd like to see the CSS / JS parsers used by that module not pulled in automatically.  Really I could see identity-map being split into two modules, one for js (acorn) and one for css.  This would make it easy to replace the parsers without forcing people to switch.

I feel similarly about having the main module not directly depend on `@gulp-sourcemaps/map-sources` though not as strongly since that plugin doesn't pull any JS/CSS parsers

I realize that not pulling these plugins would be another breaking change but I still felt the need to mention it.

**Comment by phated** [comment]
That's very much the goal.  It'll be really nice when people can just grab the modules they need and plug them together.

**Comment by coreyfarrell** [comment]
@phated How would you feel about updating the `gulp-sourcemaps` README so it tells people to use those modules directly instead of the `mapSources` and `identityMap` functions?

**Comment by phated** [comment]
@coreyfarrell when I've had a chance to implement the individual modules, I then import them as an export of this "grab bag" module, add an example and deprecate the option in the documentation.  Do you think that's not enough?  I was hoping for a gentle deprecation path before we rip out all the options in the major.

**Comment by coreyfarrell** [comment]
@phated My goal is to be able to install bare essentials, pick the pieces I need.  What if we create new modules `@gulp-sourcemaps/init` and `@gulp-sourcemaps/write`? We could eventually make it so `gulp-sourcemaps` doesn't have any actual code, just an index.js to import / re-export the default plugins?  I say eventually because I don't think the new `init` and `write` modules should ever have the deprecated options, so `gulp-sourcemaps` wouldn't be able to use them until it's time to actually remove those options.

**Comment by phated** [comment]
I agree but gulp 4 ships with init and write - that was the motivation behind this.

**Comment by coreyfarrell** [comment]
I'm sorry I'm not sure what you mean?

**Comment by phated** [comment]
@coreyfarrell https://github.com/gulpjs/vinyl-fs#optionssourcemaps

**Comment by coreyfarrell** [comment]
Ah I can just get the init / write functionality of gulp4 (vinyl-fs) and combine with the `map-sources` or whichever plugins I need.  Forget everything I said about wanting to reduce the dependencies of `gulp-sourcemaps` itself.

**Comment by coreyfarrell** [comment]
@phated Could we move https://github.com/gulpjs/vinyl-fs/blob/master/lib/dest/sourcemap.js into a stand-alone `@gulp-sourcemap/write` module?  My goal is to take my stream of vinyl-fs objects, run sourcemapsWrite to add `sourceMappingURL` comment / create vinyl-fs objects for the sourcemaps, then perform more steps before calling `gulp.dest`.  The issue with the `gulp@4` method of writing sourcemaps is that I cannot do anything between adding the sourcemaps to the stream and writing to disk (gulp.dest does both).

My use case is to support [nginx static gzip module](https://nginx.org/en/docs/http/ngx_http_gzip_static_module.html).  Given a gulp stream with `index.js` that has a sourcemap, the last argument of my first `pump` is a call to `sourcemapWrite`.  I pump the result of that into `gulp-clone` then `gulp-gzip` before merging with the original stream and pumping the merged stream to `gulp.dest` (where I do not enable writing sourcemaps).

I can take the lead on creation of this new module (including tests with full coverage of course).  I just don't want to start work without getting a thumbs up on the idea of separating it.

**Comment by phated** [comment]
@coreyfarrell I think your use case makes sense but let me confirm my understanding and then make another suggestion.

So you want to have the logic of "plucking" the sourcemap from the original `file.contents` and constructing a vinyl object from that sourcemap content, while rewriting the `sourceMappingURL` comment.  Then pushing both of those vinyl objects onto the stream, without writing anything to the filesystem.  Is that correct?

I don't think the exact same implementation from vinyl-fs belongs in the new module, because it would leak some weird features of vinyl-fs (like resolving all function options, bleh); However, we need an `@gulp-sourcemaps` module that lets us deprecate the `sourceMappingURL` and `sourceMappingURLPrefix` options of this module.  Do you think those options could be bundled into the new module?  A possible name could be `@gulp-sourcemaps/extract-sourcemap`

**Comment by coreyfarrell** [comment]
I want the logic of adding a new vinyl-fs to the stream based on `file.sourceMap` and adding the `sourceMappingURL` comment to `file.contents`.

I created a demo with a minimum gulpfile.js showing a cut down version of exactly what I'm doing at https://github.com/coreyfarrell/gulp-sourcemap-demo.  I agree it would be nice if I could avoid pretending to provide a `resolve-options` object.  I was thinking that the `@gulp-sourcemaps/write` function would accept multiple types for the argument - `resolve-options` for the sake of `gulp.dest` or scalar (string or boolean).

I'm not sure how `sourceMappingURL` and `sourceMappingURLPrefix` would be reimplemented, I think this would have to be added to the `write` function of the `vinyl-sourcemap` module first?  I'm not looking to create a bunch of duplicate code so `vinyl-sourcemap` seems like the right place.

**Comment by phated** [comment]
@coreyfarrell why not just use the first `gulp.dest` as a passthrough?  Let me see if I can't submit a PR to your sample.

**Comment by coreyfarrell** [comment]
Oh wow, I didn't realize that `gulp.dest` returned the stream intact.  This works great and is much simpler code than what I was doing.  Also thanks for pointing out `require('stream').pipeline`.  In my own projects I only work with latest LTS so this is available to me, I'll be able to drop the `pump` dependency.
